[Reader Mode] Add theme support features from `theme.json`

thelovekesh opened the pull request:
## Summary

Implemented in #6042, reader themes provide support for producing color palettes, gradient presets, and font size styles. This feature leverages `get theme support` to identify the necessary presets for generating the styles. 

Although `theme.json` can be used by any theme, it serves as a foundation for the new block themes. This PR seeks to enable theme support features styles generation using the `theme.json` presets to address the block styles breaking issues in reader mode.

Fixes #7471

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

thelovekesh committed:
Add theme support features from theme.json

thelovekesh commented:
@westonruter Although the support for `theme.json` was added in WP 5.8, the required APIs like `wp_get_global_settings` was added in WP 5.9. Should I look for a workaround to get it worked with WP 5.8 as well or we are good to support it on WP 5.9+?

---
1. Right now our CSS unit measurement doesn't look robust as it only handles all values in `px`

https://github.com/ampproject/amp-wp/blob/2a3d437a50290c93359f50c54c56f55b87313df6/src/ReaderThemeSupportFeatures.php#L421

but there might be cases when someone defines them in `em` or `rem`.

2. Also, we will need to calculate values for `clamp()` to give support for fluid typography. 

To solve the above scenarios WP 6.1 has two functions:
- [`wp_get_typography_value_and_unit()`](https://github.com/WordPress/WordPress/blob/9caf1c4adeddff2577c24d622ebbbf278a671271/wp-includes/block-supports/typography.php#L297)
- [`wp_get_computed_fluid_typography_value()`](https://github.com/WordPress/WordPress/blob/9caf1c4adeddff2577c24d622ebbbf278a671271/wp-includes/block-supports/typography.php#L385)

but the caveat is they are available prior to WP 6.1.  

I'm considering defining them in our own class right now. Please let me know your thoughts?

westonruter commented:
If someone is using block themes, I'd say it's highly unlikely that they would be holding off on updating WordPress core. Therefore, if you want to require WordPress 6.1+ to fully support fluid typography, that's fine with me.

thelovekesh commented:
But I think we should update the way of determining measurement units. Someone can use different measurement unit even when fluid typography is not being used.

thelovekesh committed:
Add `ReaderThemeSupportFeatures::get_typography_value_and_unit()`

- This is taken from `wp_get_typography_value_and_unit()`

thelovekesh committed:
Add support for fluid typography in reader mode

thelovekesh committed:
Avoid reducing theme support features for fontSizes while using theme.json

thelovekesh committed:
Add check to return correct font size format

thelovekesh committed:
Add test cases for `ReaderThemeSupportFeatures::maybe_use_theme_json()`

thelovekesh commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -385,4 +490,90 @@ public function get_relative_luminance_from_hex( $hex ) {
+	 * }
+	 * @return string|null The value and unit, or null if the value is empty.
+	 */
+	private function get_typography_value_and_unit( $raw_value, $options = [] ) {
```
We can ignore coverage for this copied method?

westonruter replied:
Sure

thelovekesh committed:
Update `ReaderThemeSupportFeatures::maybe_use_theme_json()` to be skipped on WP < 5.9

thelovekesh commented:
PR can be tested for fluid typography by disabling CSS tree-shaking till #7291 get fixed.

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -327,10 +422,20 @@ private function print_editor_font_sizes_styles( array $font_sizes ) {
 				sanitize_key( $font_size[ self::KEY_SLUG ] ),
-				(float) $font_size[ self::KEY_SIZE ]
+				function_exists( 'wp_get_typography_font_size_value' )
+					? esc_attr( wp_get_typography_font_size_value( $font_size ) )
+					: esc_attr( $font_size[ self::KEY_SIZE ] )
```
Minor point, but I don't think `esc_attr()` is right here. I think `strip_tags()` is the right choice. See https://github.com/WordPress/wordpress-develop/blob/200868214a1ae0a108dac491677ba82e7541fc8d/src/wp-includes/theme.php#L1891-L1896

thelovekesh replied:
It does make sense, but I am not sure if we will ever get tags here. Or better to never assume and update it to use `strip_tags()` 🤔

westonruter replied:
Better to never assume, yeah

thelovekesh replied:
Just realized, we will still need the escaping, as `strip_tags` is not an escaping utility.

westonruter replied:
Well, it is somewhat. It's what core uses in `wp_custom_css_cb()`.

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -385,4 +490,90 @@ public function get_relative_luminance_from_hex( $hex ) {
+			_doing_it_wrong(
+				__METHOD__,
+				esc_html__( 'Raw size value must be a string, integer, or float.', 'default' ),
+				'2.4.1'
```
Shouldn't this actually be the version that is in core?

westonruter replied:
Well, it's not in core. Nevermind.

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -219,9 +294,29 @@ public function get_theme_support_features( $reduced = false ) {
+
+		// Do not rely on `wp_is_block_theme()` as theme.json can be used in non-block themes.
+		// See: https://developer.wordpress.org/themes/advanced-topics/theme-json/#a-theme-json-can-be-added-to-any-theme.
+		return ( is_readable( get_stylesheet_directory() . '/theme.json' ) ? true : is_readable( get_template_directory() . '/theme.json' ) );
```
This would seem simpler as:

```suggestion
		return is_readable( get_stylesheet_directory() . '/theme.json' ) || is_readable( get_template_directory() . '/theme.json' );
```

westonruter replied:
Actually, I suggest using `wp_theme_has_theme_json()` if it exists, or else copy the function as a method into this class:

https://github.com/WordPress/wordpress-develop/blob/200868214a1ae0a108dac491677ba82e7541fc8d/src/wp-includes/global-styles-and-settings.php#L384-L414

Note how it caches the file exists check.

thelovekesh replied:
> Note how it caches the file exists check.

Oh yes. Bette to copy it since we will need to check for theme.json on WP 5.9+ and this function is added in WP 6.2.

thelovekesh committed:
Merge branch 'develop' into fix/reader-themes/block-theme-styles

thelovekesh committed:
Update checks for theme.json with `theme_has_theme_json`

- Copied from `wp_theme_has_theme_json`

thelovekesh committed:
Revert test cases for `ReaderThemeSupportFeatures::maybe_use_theme_json()`

This reverts commit 1fc0531ff9c7721088d8cb87c418d8df1d517339.

thelovekesh committed:
Update `::theme_has_theme_json()` to bail if `wp_get_global_settings()` not exists

thelovekesh committed:
Ignore code coverage for functions taken from WP Core

thelovekesh committed:
Update escaping to use `wp_strip_all_tags()` for css values

thelovekesh committed:
Add tests to verify when fontSize is null in ::get_typography_value_and_unit()

thelovekesh committed:
Add tests for getting theme support features from theme.json

thelovekesh committed:
Skip tests theme.json theme features on WP < 5.9

thelovekesh committed:
Merge branch 'develop' into fix/reader-themes/block-theme-styles

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -203,13 +256,35 @@ public function update_cached_theme_support() {
 	 */
 	public function get_theme_support_features( $reduced = false ) {
```
There is missing test coverage in this method. Is that feasible to add?

thelovekesh replied:
Tests are there but they were wrongfully skipped due to https://github.com/ampproject/amp-wp/pull/7481/commits/f1d16af946f18b3f2e9d44fb42e7724be265ffa3.

I have fixed it in https://github.com/ampproject/amp-wp/pull/7481/commits/722476c6b1c61b26b08eda42d954a95e115398c5

thelovekesh committed:
Fix test skip conditiona for `::test_get_theme_support_features_with_theme_json()`

thelovekesh committed:
Merge branch 'develop' into fix/reader-themes/block-theme-styles

thelovekesh committed:
Remove React 17 requirement for `AfterActivationSiteScan` service

- Why? React 17 code will continue to work with React 18 by invoking a legacy API

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -327,10 +403,20 @@ private function print_editor_font_sizes_styles( array $font_sizes ) {
 				sanitize_key( $font_size[ self::KEY_SLUG ] ),
-				(float) $font_size[ self::KEY_SIZE ]
+				function_exists( 'wp_get_typography_font_size_value' )
+					? esc_attr( wp_strip_all_tags( wp_get_typography_font_size_value( $font_size ) ) )
+					: esc_attr( wp_strip_all_tags( $font_size[ self::KEY_SIZE ] ) )
```
This can just use `strip_tags()` since `esc_attr()`  could introduce entities which break CSS.

```suggestion
					? strip_tags( wp_get_typography_font_size_value( $font_size ) )
					: strip_tags( $font_size[ self::KEY_SIZE ] )
```

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -385,4 +471,145 @@ public function get_relative_luminance_from_hex( $hex ) {
+	 * @return bool False if `wp_get_global_settings()` not exists or theme.json not found, true otherwise.
+	 */
+	private function theme_has_theme_json() {
+		// wp_get_global_settings() is only available in WP 5.9+.
```
This should defer to `wp_theme_has_theme_json()` if it exists:

```suggestion
		if ( function_exists( 'wp_theme_has_theme_json' ) ) {
			return wp_theme_has_theme_json();
		}
		// wp_get_global_settings() is only available in WP 5.9+.
```

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -385,4 +471,145 @@ public function get_relative_luminance_from_hex( $hex ) {
+	 * @return string|null The value and unit, or null if the value is empty.
+	 */
+	private function get_typography_value_and_unit( $raw_value, $options = [] ) {
+		if ( ! is_string( $raw_value ) && ! is_int( $raw_value ) && ! is_float( $raw_value ) ) {
```
```suggestion
		if ( function_exists( 'wp_get_typography_value_and_unit' ) ) {
			return wp_get_typography_value_and_unit( $raw_value, $options );
		}
		if ( ! is_string( $raw_value ) && ! is_int( $raw_value ) && ! is_float( $raw_value ) ) {
```

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -203,13 +256,35 @@ public function update_cached_theme_support() {
+				$feature_value   = [];
+				$global_settings = wp_get_global_settings( self::SUPPORTED_THEME_JSON_FEATURES[ $feature_key ], 'theme' );
+
+				if ( isset( $global_settings[ self::KEY_THEME ] ) ) {
+					$feature_value = array_merge( $feature_value, $global_settings[ self::KEY_THEME ] );
+				}
+
+				if ( isset( $global_settings[ self::KEY_DEFAULT ] ) ) {
+					$feature_value = array_merge( $feature_value, $global_settings[ self::KEY_DEFAULT ] );
+				}
```
I confess I am not up to speed on `theme.json`, but wouldn't the `theme` value supercede the `default` value? In other words, shouldn't these two conditions be swapped? (I'd be grateful to be educated otherwise!)

thelovekesh replied:
The following is the output (in array) when we use `wp_get_global_settings` to retrieve theme features from `theme.json`:

Here, `theme` is derived from `theme.json`, and the `default` preset is coming from the block editor.

<details><summary>Theme Features</summary>

```json
{
  "default": [
    {
      "name": "Black",
      "slug": "black",
      "color": "#000000"
    },
    {
      "name": "Cyan bluish gray",
      "slug": "cyan-bluish-gray",
      "color": "#abb8c3"
    },
    {
      "name": "White",
      "slug": "white",
      "color": "#ffffff"
    },
    {
      "name": "Pale pink",
      "slug": "pale-pink",
      "color": "#f78da7"
    },
    {
      "name": "Vivid red",
      "slug": "vivid-red",
      "color": "#cf2e2e"
    },
    {
      "name": "Luminous vivid orange",
      "slug": "luminous-vivid-orange",
      "color": "#ff6900"
    },
    {
      "name": "Luminous vivid amber",
      "slug": "luminous-vivid-amber",
      "color": "#fcb900"
    },
    {
      "name": "Light green cyan",
      "slug": "light-green-cyan",
      "color": "#7bdcb5"
    },
    {
      "name": "Vivid green cyan",
      "slug": "vivid-green-cyan",
      "color": "#00d084"
    },
    {
      "name": "Pale cyan blue",
      "slug": "pale-cyan-blue",
      "color": "#8ed1fc"
    },
    {
      "name": "Vivid cyan blue",
      "slug": "vivid-cyan-blue",
      "color": "#0693e3"
    },
    {
      "name": "Vivid purple",
      "slug": "vivid-purple",
      "color": "#9b51e0"
    }
  ],
  "theme": [
    {
      "color": "#ffffff",
      "name": "Base",
      "slug": "base"
    },
    {
      "color": "#000000",
      "name": "Contrast",
      "slug": "contrast"
    },
    {
      "color": "#9DFF20",
      "name": "Primary",
      "slug": "primary"
    },
    {
      "color": "#345C00",
      "name": "Secondary",
      "slug": "secondary"
    },
    {
      "color": "#F6F6F6",
      "name": "Tertiary",
      "slug": "tertiary"
    }
  ]
}
```

</details> 

<details><summary>Block Editor View</summary>
<p>

![image](https://user-images.githubusercontent.com/54371619/225860752-6f4eacdb-0b9c-4cb7-9be9-be40561c6d14.png)

</p>
</details> 

Therefore I believe they do not intend for one to supersede the other. The user may select from the `theme` preset or `default` preset at will.

westonruter replied:
Ah, it is not merging associative arrays, so my concern is invalid.

thelovekesh committed:
Use `strip_tags()` to escape css

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Use `wp_theme_has_theme_json()` if exists

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Use `wp_get_typography_value_and_unit` if exists

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Fix phpcs errors due to `strip_tags()`

thelovekesh committed:
Fix usage of `::get_typography_value_and_unit()`

thelovekesh committed:
Merge branch 'develop' into fix/reader-themes/block-theme-styles

thelovekesh committed:
Add test case to verify fluid typography in reader mode

thelovekesh committed:
Fix test skip message

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -203,13 +256,35 @@ public function update_cached_theme_support() {
+
 		foreach ( array_keys( self::SUPPORTED_FEATURES ) as $feature_key ) {
-			$feature_value = current( (array) get_theme_support( $feature_key ) );
+			if ( $this->theme_has_theme_json() ) {
+				$feature_value   = [];
+				$global_settings = wp_get_global_settings( self::SUPPORTED_THEME_JSON_FEATURES[ $feature_key ], 'theme' );
```
While the `theme_has_theme_json` method has a check for `wp_get_global_settings`, it also now short-circuits if the `wp_theme_has_theme_json` function exists. Therefore, I think it may be safer just to move the function exists check here:

```suggestion
			if ( $this->theme_has_theme_json() && function_exists( 'wp_get_global_settings' ) ) {
				$feature_value   = [];
				$global_settings = wp_get_global_settings( self::SUPPORTED_THEME_JSON_FEATURES[ $feature_key ], 'theme' );
```

thelovekesh replied:
Yes, it seems much better.

`wp_theme_has_theme_json` is introduced in WP 6.2 so `wp_get_global_settings` will exists, but this solution looks cleaner to me.

westonruter commented on src/ReaderThemeSupportFeatures.php:
```diff
@@ -385,4 +480,156 @@ public function get_relative_luminance_from_hex( $hex ) {
+		if ( function_exists( 'wp_theme_has_theme_json' ) ) {
+			return wp_theme_has_theme_json();
+		}
+
+		// wp_get_global_settings() is only available in WP 5.9+.
+		if ( ! function_exists( 'wp_get_global_settings' ) ) {
+			return false;
+		}
```
See above.

```suggestion
```

westonruter replied:
This then makes the function an actual 1:1 copy of `wp_theme_has_theme_json()`

westonruter commented on tests/php/src/ReaderThemeSupportFeaturesTest.php:
```diff
@@ -303,6 +316,60 @@ public function test_get_theme_support_features() {
+			$features['editor-font-sizes']
+		);
+
+		switch_theme( $current_theme->get_stylesheet() );
```
This shouldn't be necessary since the test tear-down handles this automatically.

```suggestion
```

westonruter commented on tests/php/src/ReaderThemeSupportFeaturesTest.php:
```diff
@@ -403,6 +470,22 @@ public function test_print_theme_support_styles_reader( $is_legacy ) {
+			$this->assertStringContainsString( 'font-size: clamp(', $output );
+			$this->assertStringContainsString( '+ ((', $output );
+			$this->assertStringContainsString( ':root .has-small-font-size { font-size: clamp(0.875rem, 0.875rem + ((1vw - 0.48rem) * 0.24), 1rem); }', $output );
+
+			switch_theme( $current_theme->get_stylesheet() );
```
```suggestion
```

thelovekesh committed:
Remove resetting active theme state in test method

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Remove `wp_get_global_settings` function exists check from `::theme_has_theme_json()`

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter commented:
This seems very close. I'm testing with 2023 with the following content:

```html
<!-- wp:paragraph {"fontSize":"xx-large"} -->
<p class="has-xx-large-font-size">Hello!</p>
<!-- /wp:paragraph -->
```

Primary Theme (2023) | Reader Theme (2021) Before | Reader Theme (2021) After
---------------------|-----------------------------|--------------------------
![image](https://user-images.githubusercontent.com/134745/225982910-601cd5c0-a321-47d9-acaa-24b4cef524b1.png) | ![image](https://user-images.githubusercontent.com/134745/225982857-d69100f5-5345-44c3-ade8-e193d55c99ed.png) | ![image](https://user-images.githubusercontent.com/134745/225982983-28ed0486-95b2-4a2a-a98a-a8a8004e3e04.png)

So that's great. What's not quite there yet is the Legacy theme:

> ![image](https://user-images.githubusercontent.com/134745/225983280-3c65f7f2-e431-4bb4-85de-fbad57448ae6.png)

Something's not right with margin or line-height or something.

thelovekesh commented:
Seems like on legacy themes `line-height` is causing issues.

<details><summary>With <code>line-height: 1.75em</code></summary>

![image](https://user-images.githubusercontent.com/54371619/225984283-5cdf7637-4238-411d-b0b8-c0fb4199e19f.png)
</details> 

<details><summary>With <code>line-height: 1.5</code></summary>

![image](https://user-images.githubusercontent.com/54371619/225984935-ca7219c3-c539-4fc3-b240-f80b47f48726.png)
</details>

westonruter commented:
In TT3, there is this rule applying to the content container:

```css
body .is-layout-flow > * + * {
    margin-block-start: 1.5rem;
    margin-block-end: 0;
}
```

Applying these styles to `.amp-wp-article-content` mostly fixes the issue:

![image](https://user-images.githubusercontent.com/134745/225985633-ed6d5356-cc6f-427d-854e-a343f40b9482.png)

Not perfectly, but better.

westonruter commented:
OK, let's change this:

https://github.com/ampproject/amp-wp/blob/4337065a80df965a4d49d2858e6fcced339b90bc/templates/style.php#L116

To just `1.75`.

westonruter commented:
`line-height: 1.75em` | `line-height: 1.75`
---------|--------
![image](https://user-images.githubusercontent.com/134745/225988043-8e9dc610-77ab-455b-b46e-679bc24d65a3.png) | ![image](https://user-images.githubusercontent.com/134745/225988093-690b6d24-3a4a-4d49-8574-dd3370a9d28d.png)

westonruter commented:
Something else that isn't quite right, even with `line-height:1.75`. Adding a background color to a paragraph that has XXL font size:

```html
<!-- wp:paragraph {"backgroundColor":"primary","fontSize":"xx-large"} -->
<p class="has-primary-background-color has-background has-xx-large-font-size">Hello!</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent interdum, elit non cursus tincidunt, nibh nibh congue metus, ac gravida nunc ipsum ut augue. Donec tempor pulvinar maximus. Sed et tellus nec sapien vulputate sollicitudin. Nam auctor nulla imperdiet, sodales leo ultricies, cursus metus. Aenean ullamcorper dui vel neque luctus faucibus non vel odio. Aenean convallis mattis nisi, nec fringilla sem sagittis eget. Duis ut quam venenatis, blandit elit id, iaculis ipsum. Curabitur tristique euismod nisl, a finibus massa dignissim ac. Nullam nisl orci, imperdiet at libero sed, pulvinar vehicula lacus. Duis eleifend sollicitudin pharetra. Maecenas sapien nunc, mollis a velit sed, ultricies accumsan tortor. Ut accumsan sapien id turpis eleifend, at eleifend orci volutpat. Etiam massa arcu, congue sit amet sem vitae, gravida efficitur sem. Curabitur ornare velit velit, ut accumsan tellus convallis at. Nunc sagittis nisl mauris, nec hendrerit leo sagittis et.</p>
<!-- /wp:paragraph -->
```

TT3 | TT2 Reader | Legacy Reader
-----|-----------|--------------
![image](https://user-images.githubusercontent.com/134745/225991358-a4608004-79c3-4adb-94ef-72457f4f9e54.png) | ![image](https://user-images.githubusercontent.com/134745/225991276-91e65653-41e5-465a-bb64-2f0accc10aff.png) | ![image](https://user-images.githubusercontent.com/134745/225991170-2b66d84c-0e42-4a03-adc5-7c4bdc7e6160.png)

thelovekesh commented:
Seems like the `xx-large` font size requires more width on the screen but we have restricted it to `840px` in `.amp-wp-article` which wraps the whole article in it.

https://github.com/ampproject/amp-wp/blob/d03e17921daab6f614d463d8274f8154a6f20972/templates/style.php#L203

westonruter commented:
Oh, if I actually look at Legacy in a mobile viewport it renders fine:

[Screen recording 2023-03-17 13.20.20.webm](https://user-images.githubusercontent.com/134745/226028786-3d3cec77-d710-4a4e-92a2-0c15c5789ce1.webm)

We could fix this by adding negative margins:

```css
p.has-background {
    margin-left: -2.375em;
    margin-right: -2.375em;
}
```

[Screen recording 2023-03-17 13.24.26.webm](https://user-images.githubusercontent.com/134745/226030205-1257d667-ed52-40a3-a17e-0474ccf8bf02.webm)

westonruter commented:
Another case to check for: groups with backgrounds.

```html
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|40","right":"var:preset|spacing|40","bottom":"var:preset|spacing|40","left":"var:preset|spacing|40"}}},"backgroundColor":"primary","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-primary-background-color has-background" style="padding-top:var(--wp--preset--spacing--40);padding-right:var(--wp--preset--spacing--40);padding-bottom:var(--wp--preset--spacing--40);padding-left:var(--wp--preset--spacing--40)"><!-- wp:paragraph {"backgroundColor":"luminous-vivid-amber"} -->
<p class="has-luminous-vivid-amber-background-color has-background">First</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"backgroundColor":"vivid-red"} -->
<p class="has-vivid-red-background-color has-background">Second</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

Should look like this:

![image](https://user-images.githubusercontent.com/134745/226039974-f7d0f2b2-9a38-445d-8092-06f388a88a35.png)

thelovekesh committed:
Update line height to `1.75` in legacy reader theme

thelovekesh committed:
Fix background margin; block group margin in reader mode

thelovekesh committed:
Add `::print_spacing_sizes_custom_properties()`

- Print spacing sizes custom properties like `-wp--preset--spacing--30: clamp(1.5rem,5vw,2rem);`).

thelovekesh committed:
Update global settings keys with constants

thelovekesh committed:
Add test cases to account for spacing size features in `theme.json`

thelovekesh committed:
Update `::print_spacing_sizes_custom_properties()` to determine correct spacing preset

thelovekesh committed:
Update PHPUnit multisite workflow to continue on error

thelovekesh committed:
Merge branch 'develop' into fix/reader-themes/block-theme-styles

westonruter commented on .github/workflows/build-test-measure.yml:
```diff
@@ -577,6 +577,7 @@ jobs:
     needs: pre-run
     runs-on: ubuntu-latest
     if: needs.pre-run.outputs.changed-php-count > 0
+    continue-on-error: true
```
Why specifically?

thelovekesh replied:
Sometimes external-http tests may fail the workflow, causing main workflow to exit early.

westonruter commented:
Looking good!

Non-AMP | Desktop Legacy AMP | Mobile Legacy AMP
--|--|--
![image](https://user-images.githubusercontent.com/134745/226440319-d15fc5cf-a800-40c7-8c0c-db0a4b0c9b19.png) | ![image](https://user-images.githubusercontent.com/134745/226440397-10a65c5a-f91b-4529-933c-fe5b24392d09.png) | ![image](https://user-images.githubusercontent.com/134745/226440567-266f8495-7b81-460a-895f-3b8568f7fb30.png)
